django/db/models/query.py

```diff
@@ -654,19 +654,20 @@ def _earliest(self, *fields):
                 "earliest() and latest() require either fields as positional "
                 "arguments or 'get_latest_by' in the model's Meta."
             )
-
-        assert not self.query.is_sliced, \
-            "Cannot change a query once a slice has been taken."
         obj = self._chain()
         obj.query.set_limits(high=1)
         obj.query.clear_ordering(force_empty=True)
         obj.query.add_ordering(*order_by)
         return obj.get()
 
     def earliest(self, *fields):
+        if self.query.is_sliced:
+            raise TypeError('Cannot change a query once a slice has been taken.')
         return self._earliest(*fields)
 
     def latest(self, *fields):
+        if self.query.is_sliced:
+            raise TypeError('Cannot change a query once a slice has been taken.')
```

Comment:
`Cannot change a query once a slice has been taken.`  Repeated many times, is it possible to extract a variable.

Reply:
IMO an extra hook or variable is not necessary.